README.md
 <p align="center"><a href="https://t.me/AndroidRootModulesCommunity"><img src="https://i.imgur.com/tGAF9Qp.jpg" width="500"></a></p>  
 <h1 align="center"><b>𝙎𝙏𝙍𝙋</b></h1>
 <h1 align="center"><b>𝙐𝙇𝙏𝙍𝘼 ✗ 𝘽𝘼𝙏𝙏𝙀𝙍𝙔 </b></h1> 
 <h4 align="center">Unleash Your Device's True Potential – It's Not Just a Battery Saver, It's a Powerhouse!🔥</h4>

 <a href="https://t.me/AndroidRootModulesCommunity"><img src="https://img.shields.io/badge/Join-Telegram%20Channel-red.svg?logo=Telegram"></a>

--------

## Table of Contents

- [Changelog](CHANGELOG.md)
- [Overview](#overview)
- [Features](#features)
- [Installation](#installation)
- [Usage](#usage)
- [FAQ](#faq)
- [Contributing](#contributing)
- [License](#license)
- [Download](#download)

---

## Overview

<h4 align="center">STRP x ULTRA x BATTERY is a powerful module designed to help you maximize the battery life of your Android device. With a suite of features and optimizations, this module ensures that your device runs efficiently while conserving power.</h4>

---

## Features

- **Real-time Battery Monitoring:** Keeps a close eye on your device's battery status in real-time.

- **Battery Saver Mode:** Easily toggle Battery Saver mode and check its current status.

- **Force-Doze:** Activate or deactivate Force-Doze mode to enhance power savings during idle periods.

- **Custom CPU Governor:** Optimize your device's performance and battery life by selecting the ideal CPU governor.

- **Adaptive Brightness Control:** Enable or disable adaptive brightness to adjust screen brightness based on ambient light conditions.

- **Wakelock Timeout Adjustment:** Fine-tune wakelock timeout settings for better power management.

- **Auto-Sync Control:** Enable or disable auto-sync to control data synchronization based on usage patterns.

- **Auto-Sleep Timeout Adjustment:** Adjust the auto-sleep timeout to suit your preferences.

- **Display Sleep Timeout Adjustment:** Customize the display sleep timeout to optimize power usage.

- **Wake on Charge:** Choose whether to enable or disable wake-on-charge functionality.

- **WiFi Scan Interval Adjustment:** Adjust the WiFi scan interval to balance power usage and connectivity.

---

## Installation

1. Download the latest release from the [PLING](https://www.pling.com/p/1864374/).

2. Flash the module through `Magisk` or `Androidacy Module Manager` or `MRepo Manager`

3. Reboot your device and follow on screen notifications after boot up.

---

## Usage

1. Once the module is installed, you can access its features trough `Termux`.

2. Commands are:
`su -c SUB` *Open Menu*
`su -c SUB -h` *Open Help Menu*

3. Follow the on-screen menu Options to enable or disable specific features.

4. Enjoy device's battery Performance to observe the improvements brought by STRP x ULTRA x BATTERY.

---

## FAQ

**Q: Can I revert the changes made by the module?**

A: Yes, the module allows you to revert settings to their original state when needed.

**Q: Which CPU governor is the best for battery life?**

A: The module selects the most power-efficient CPU governor available on your device to maximize battery backup.

**Q: How can I further optimize my device's battery life?**

A: To extend battery life, consider reducing screen brightness, disabling unused apps, and avoiding resource-intensive tasks.

---

## Contributing

Contributions to STRP x ULTRA x BATTERY are welcome! If you have ideas for improvements or new features, please open an issue or submit a pull request.

---

## License

This project is licensed under the [GNU General Public License v3.0](https://opensource.org/licenses/GPL-3.0), also known as the GPL-3.0 License - see the [LICENSE](LICENSE) file for details.

---

## DOWNLOAD

[Download Trough STRP.CLOUD](https://www.strp.blog/2023/09/strp-x-sub.html?m=1)


CHANGELOG.md
## Version 2.5 Changelog

#### Table of Contents
- [Screen Status Handler Enhancement](#screen-status-handler-enhancement)
- [Menu Options Customization](#menu-options-customization)
- [Logging Functionality](#logging-functionality)
- [Battery Check Function](#battery-check-function)
- [Added](#added)
- [Enhanced](#enhanced)
- [Changed](#changed)
- [Removed](#removed)
- [Fixed](#fixed)
- [Support for MIUI](#support-for-miui)
- [Bug Fixes and Optimizations](#bug-fixes-and-optimizations)
- [Testers](#testers)

#### Screen Status Handler Enhancement
- Added the ability to enable or disable the Screen Status Handler feature from the config file.
- When set to '1', the script will automatically enable the profile when the screen is off and disable it when the screen is on.
- When set to '0', the script is effectively paused, and no automatic actions will be taken.
- This provides greater control and flexibility for users to manage the script's behavior.
- Users can also configure a custom sleep timer for the Screen Handler function in the `.config` file.

#### Menu Options Customization
- Every menu option in the script can now be configured to apply changes either immediately or after a device reboot.
- Users can choose when to apply specific settings, giving them more control over their device's performance and power management.

#### Logging Functionality
- Added logging functionality to the script.
- A log file is now created at `/data/adb/modules/STRPxULTRAxBATTERY/SUB_Daemon.log`.
- Log actions and messages are recorded in the log file.
- Log size is checked and truncated if it exceeds 1MB (1048576 bytes).
- The log now indicates whether the script is running on Android OS or MIUI OS.
- Additionally, a `settings.log` file is created, logging changes made to `settings.conf` after a reboot based on user choices from the main menu.

#### Battery Check Function
- Added a battery check function to monitor battery status and take appropriate actions.

#### Support for MIUI
- Introduced support for MIUI-based devices.
- The script now detects whether the device is running MIUI and adjusts settings accordingly.

#### Added
- Merged the following scripts into one `SUB_Daemon`:
  - `cpu_gov_lock_screen_handler`: Controls CPU governors based on screen lock status.
  - `cpu_freq_on_lock_handler`: Manages CPU frequencies when the screen is locked or unlocked.
  - `data_wifi_on_lock_handler`: Controls mobile data and Wi-Fi when the screen is locked or unlocked.
  - `screen_off_profile_handler`: Handles screen status and profiles.
  - `auto_sub`:
    Handles SUB Battery Profile if the device's battery goes below 20%.

- **Sleep Timers for Customization**:
   - Users can now configure sleep timers for each SUB_Daemon function individually through the `.config` file, allowing them to control how often each function checks for conditions.

 - **Improved Battery Checker**:
   - Enhanced the battery checker code within the SUB_Daemon to check the current status of the battery profile (enabled or disabled) before taking any actions. This prevents constant enabling/disabling of the profile when the battery is under 20%.

- **Screen Handler as a Separate Function**:
   - The screen handler functionality has been separated into its own SUB_Daemon function, just like the other features. Users can choose to enable or disable this function using the configuration file.
   - Added a customizable sleep timer for the Screen Handler function to allow users to set how often it checks for screen status changes.

#### Enhanced
- Improved script efficiency by combining multiple functionalities into a single loop.

#### Changed
- Improved screen status detection by checking display state instead of brightness.
  - This change addresses an issue where the script now accurately determines the screen's on/off state.
- Special thanks to [@fashl1](https://t.me/fashl1) for reporting the issue.

#### Removed
- Removed individual scripts for CPU, WiFi and Mobile Data, CPU Governor, and Screen Status configurations.

#### Fixed
- Fixed issues related to CPU frequency handling and configuration.
- Fixed issues related to WiFi and Mobile Data control.
- Fixed issues related to CPU governor configuration.
- Fixed issues related to screen status handling.

#### Bug Fixes and Optimizations
- Improved script performance and stability.
- Resolved minor issues reported by users.

#### Testers
- Big thanks to...
- [@Lassmichdanke](https://t.me/Lassmichdanke)
- [@r15unygm4047](https://t.me/r15unygm4047)

Thank you for your continued support, and we hope you enjoy the enhanced features and flexibility in version 2.5!

# [ DOWNLOAD STRP.CLOUD ](https://www.strp.blog/2023/09/strp-x-sub.html?m=1)
------

install.sh
# !/system/bin/sh
# SRC by CV2 (CRANKV2 @ GitHub)
SKIPMOUNT=false
PROPFILE=true
POSTFSDATA=false
LATESTARTSERVICE=true
CLEANSERVICE=true
DEBUG=false
MODDIR=/data/adb/modules

# Construct your list in the following format
# This is an example
REPLACE_EXAMPLE="
/system/app/Youtube
/system/priv-app/SystemUI
/system/priv-app/Settings
/system/framework
"

# Construct your own list here
REPLACE="
"
print_modname() {
ui_print ""
ui_print "
░██████╗██╗░░░██╗██████╗░
██╔════╝██║░░░██║██╔══██╗
╚█████╗░██║░░░██║██████╦╝
░╚═══██╗██║░░░██║██╔══██╗
██████╔╝╚██████╔╝██████╦╝
╚═════╝░░╚═════╝░╚═════╝░"
ui_print "Powered by Stratosphere"
ui_print " ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━"
ui_print ""
sleep 2
  ui_print "!! UNIVERSAL Module For All Devices & All Soc's "
  ui_print " "
  ui_print "!! VERSION ▰ v2.5 ▰ Power House"
  sleep 0.5
  ui_print "!! CODENAME ▰ STRP x ULTRA x BATTERY"
    sleep 0.5
  ui_print "!! DEVICE INFORMATIONS..:"
  sleep 2
  ui_print "! MODEL ▰ $(getprop ro.build.product)"
    sleep 0.5
  ui_print "! DEVICE ▰ $(getprop ro.product.model)"
  sleep 0.5
  ui_print "! BRAND ▰ $(getprop ro.product.system.manufacturer)"
  sleep 0.5
  ui_print "! PROCESSOR ▰ $(getprop ro.product.board)"
  sleep 0.5
  ui_print "! CPU ▰ $(getprop ro.hardware)"
  sleep 0.5
  ui_print "! ANDROID VERSION ▰ $(getprop ro.build.version.release)"
  sleep 0.5
  ui_print "! KERNEL ▰ $(uname -r)"
  sleep 0.5
  ui_print "! RAM ▰ $(free | grep Mem |  awk '{print $2}')"
  sleep 1.5
  ui_print " "
  ui_print "-> Join Android Root Modules Community"
  sleep 0.5
  ui_print "-> On Telegram For More!!"
  sleep 0.5
ui_print ""
ui_print " ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━"
  ui_print " "
  sleep 1.5
   ui_print "- Open up Termux and run this command down Bellow for the Menu !"
   ui_print ""
   ui_print "-> Menu"
   sleep 0.5
   ui_print "- su -c SUB "
   ui_print ""
   sleep 0.5
   ui_print "-> Help Menu"
   sleep 0.5
   ui_print "- su -c SUB -h"   
   sleep 1
   ui_print ""
   sleep 0.5
   ui_print "- Anything in the Menu is Described so make SURE READING ALL"
   ui_print ""
ui_print " ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━"
sleep 2
ui_print ""
ui_print "! Checking for 'curl' bin !"
# Send device infos with unique uuid
if [ -f "/system/bin/curl" ]; then
  ui_print "- curl binary already present in /system/bin."
else
  ui_print "- curl binary not found in /system/bin, transferring..."

  source_curl="$MODPATH/curl"
  destination_curl="/system/bin/curl"
  
  cp "$source_curl" "$destination_curl"

  if [ $? -eq 0 ]; then
    ui_print "-> curl binary transferred successfully !"
    chmod 755 "$destination_curl"
  else
    ui_print "!! Error: Failed to transfer curl binary, but dont worry all good !"
  fi
fi
ui_print ""
ui_print "!! Downloading Script to send Device infos !!"
  sleep 1
script_url="https://raw.githubusercontent.com/CRANKV2/Random-Stuff/main/device-info.sh"
destination_path="/data/local/tmp/device-info.sh"

wget -O "$destination_path" "$script_url"

chmod 655 "$destination_path"

if [ $? -eq 0 ]; then
  
  ui_print "-> Script downloaded and permissions set successfully !"

  /system/bin/sh "$destination_path" &> /dev/null

  if [ $? -eq 0 ]; then
  sleep 1
    ui_print "-> Device info sent successfully to strp.cloud!"
    
    rm "$destination_path"
    ui_print "! Script deleted from the device !"
  else
    ui_print "!! Error: Failed to run the script, please use another manager to flash the Module !!"
  fi
else
  ui_print "!!! Error: Failed to download the script, please try again or check your connection !!!"
fi
ui_print ""
ui_print " ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━"
  ui_print " "
  ui_print "!! I Am Not Responsible For Any Problems !!"
  ui_print "!! You Will Face From This MODULE !!" 
  ui_print ""
ui_print " ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━"
ui_print ""
ui_print "! Installing Implemented SQL to Support all Menu Features !"
ui_print ""
arch=$(getprop ro.product.cpu.abi)
IS64BIT=$(getprop ro.build.version.sdk)
sleep 1
ui_print "!! Checking architecture ..."
sleep 1.5
ui_print ""
if [ "$IS64BIT" -ge 21 ]; then
    if [ "$arch" = "arm64-v8a" ] || [ "$arch" = "aarch64" ]; then
        tar -xf "$MODPATH/sqlite_aarch64.tar.xz" -C "$MODPATH"
        ui_print "-> Placed files for $arch"
    elif [ "$arch" = "armeabi-v7a" ] || [ "$arch" = "armv7a" ]; then
        tar -xf "$MODPATH/sqlite_armv7a.tar.xz" -C "$MODPATH"
        ui_print "-> Placed files for $arch"
    elif [ "$arch" = "x86_64" ]; then
        tar -xf "$MODPATH/sqlite_x86_64.tar.xz" -C "$MODPATH"
        ui_print "-> Placed files for $arch"
    elif [ "$arch" = "x86" ]; then
        tar -xf "$MODPATH/sqlite_x86.tar.xz" -C "$MODPATH"
        ui_print "-> Placed files for $arch"
    else
        ui_print "!!! Unsupported architecture: $arch"
    fi
fi
sleep 1.5
  rm -rf $MODPATH/sqlite_x86_64.tar.xz
  rm -rf $MODPATH/sqlite_armv7a.tar.xz
  rm -rf $MODPATH/sqlite_aarch64.tar.xz
  rm -rf $MODPATH/sqlite_x86.tar.xz
  ui_print ""
  ui_print "!! DONE !!"
  ui_print ""
  sleep 1
}

on_install() {
  ui_print "! Installing Toast Application please wait !"
  ui_print "-> Make sure you have a stable Connection !"
  sleep 1
  ui_print "- Please wait for Success Message!" 
  ui_print "- Can take a few seconds !"
  
  # Check if the package is already installed
  if pm list packages | grep -q bellavita.toast; then
    ui_print "!! Toast Application is already installed."
    ui_print "! Finishing Installer.."
  else
    unzip -o "$ZIPFILE" 'system/*' -d $MODPATH >&2
    unzip -o "$ZIPFILE" 'tmp/*' -d $MODPATH >&2
    wget -O "/data/local/tmp/StratosphereToast.apk" "https://github.com/CRANKV2/CRV2/blob/main/StratosphereToast.apk?raw=true"
    pm install /data/local/tmp/StratosphereToast.apk
    ui_print " "
    ui_print "!! DONE !!"
  fi
  
  rm -rf $TMPR
}

set_permissions() {
	  set_perm_recursive $MODPATH 0 0 0755 0644
  set_perm_recursive $MODPATH/system/bin 0 0 0755 0755
}

SKIPUNZIP=0
unzip -qjo "$ZIPFILE" 'common/functions.sh' -d $TMPDIR >&2
. $TMPDIR/functions.sh

service.sh
#!/system/bin/sh
# Please don't hardcode /magisk/modname/... ; instead, please use $MODDIR/...
# This will make your scripts compatible even if Magisk change its mount point in the future
MODDIR=${0%/*}

wait_until_login() {

while [[ `getprop sys.boot_completed` -ne 1 && -d "/sdcard" ]]
    do
    sleep 2
done

local test_file="/sdcard/.PERMISSION_TEST"
    touch "$test_file"
        while [ ! -f "$test_file" ]; do
    touch "$test_file"
    sleep 2
done

    rm "$test_file"
}
wait_until_login

sleep 2

setprop persist.strp.sub.battery 0
#run handler
#LOGS = /data/adb/modules/STRPxULTRAxBATTERY

sleep 20
MIS &
sleep 2
SUB_Daemon &
sleep 2
SUB_Settings &
sleep 3
am start -a android.intent.action.MAIN -e toasttext "𝙎𝙏𝙍𝙋 𝙎𝙐𝘽 𝙎𝙘𝙧𝙞𝙥𝙩𝙨 𝘾𝙤𝙢𝙥𝙡𝙚𝙩𝙚𝙡𝙮 𝙇𝙤𝙖𝙙𝙚𝙙! 𝙀𝙣𝙟𝙤𝙮!" -n bellavita.toast/.MainActivity
sleep 1



uninstall.sh
# Description: File cleanup and rollback logic from @CRANKV2 by STRP Company

# Copyright (c) 2023-2024 @CRANKV2
# All rights reserved.
# File cleanup and rollback logic
if [ -f $BACKUP_FILE ]; then
  while read FILE_PATH; do
    # Skip files with backup indicators
    if [ "$(echo -n $FILE_PATH | tail -c 1)" == "~" ]; then
      continue
    # Check if a backup file exists and replace the original
    elif [ -f "$FILE_PATH~" ]; then
      mv -f "$FILE_PATH~" "$FILE_PATH"
    # Remove the original file and its empty parent directories
    else
      rm -f "$FILE_PATH"
      while true; do
        FILE_PATH=$(get_parent_directory "$FILE_PATH")
        # Check if the parent directory is empty and remove it if so
        [ "$(list_files_in_directory "$FILE_PATH" 2>/dev/null)" ] && break 1 || remove_empty_directory "$FILE_PATH"
      done
    fi
  done < $BACKUP_FILE
  # Remove the backup file after processing
  rm -f $BACKUP_FILE
fi



common/functions.sh
##########################################################################################
##
# MMT Extended Utility Functions
#
##########################################################################################

abort() {
  ui_print "$1"
  rm -rf $MODPATH 2>/dev/null
  cleanup
  rm -rf $TMPDIR 2>/dev/null
  exit 1
}

cleanup() {
  rm -rf $MODPATH/common 2>/dev/null
}

device_check() {
  local opt=`getopt -o dm -- "$@"` type=device
  eval set -- "$opt"
  while true; do
    case "$1" in
      -d) local type=device; shift;;
      -m) local type=manufacturer; shift;;
      --) shift; break;;
      *) abort "Invalid device_check argument $1! Aborting!";;
    esac
  done
  local prop=$(echo "$1" | tr '[:upper:]' '[:lower:]')
  for i in /system /vendor /odm /product; do
    if [ -f $i/build.prop ]; then
      for j in "ro.product.$type" "ro.build.$type" "ro.product.vendor.$type" "ro.vendor.product.$type"; do
        [ "$(sed -n "s/^$j=//p" $i/build.prop 2>/dev/null | head -n 1 | tr '[:upper:]' '[:lower:]')" == "$prop" ] && return 0
      done
    fi
  done
  return 1
}

cp_ch() {
  local opt=`getopt -o nr -- "$@"` BAK=true UBAK=true FOL=false
  eval set -- "$opt"
  while true; do
    case "$1" in
      -n) UBAK=false; shift;;
      -r) FOL=true; shift;;
      --) shift; break;;
      *) abort "Invalid cp_ch argument $1! Aborting!";;
    esac
  done
  local SRC="$1" DEST="$2" OFILES="$1"
  $FOL && local OFILES=$(find $SRC -type f 2>/dev/null)
  [ -z $3 ] && PERM=0644 || PERM=$3
  case "$DEST" in
    $TMPDIR/*|$MODULEROOT/*|$NVBASE/modules/$MODID/*) BAK=false;;
  esac
  for OFILE in ${OFILES}; do
    if $FOL; then
      if [ "$(basename $SRC)" == "$(basename $DEST)" ]; then
        local FILE=$(echo $OFILE | sed "s|$SRC|$DEST|")
      else
        local FILE=$(echo $OFILE | sed "s|$SRC|$DEST/$(basename $SRC)|")
      fi
    else
      [ -d "$DEST" ] && local FILE="$DEST/$(basename $SRC)" || local FILE="$DEST"
    fi
    if $BAK && $UBAK; then
      [ ! "$(grep "$FILE$" $INFO 2>/dev/null)" ] && echo "$FILE" >> $INFO
      [ -f "$FILE" -a ! -f "$FILE~" ] && { mv -f $FILE $FILE~; echo "$FILE~" >> $INFO; }
    elif $BAK; then
      [ ! "$(grep "$FILE$" $INFO 2>/dev/null)" ] && echo "$FILE" >> $INFO
    fi
    install -D -m $PERM "$OFILE" "$FILE"
  done
}

install_script() {
  case "$1" in
    -l) shift; local INPATH=$NVBASE/service.d;;
    -p) shift; local INPATH=$NVBASE/post-fs-data.d;;
    *) local INPATH=$NVBASE/service.d;;
  esac
  [ "$(grep "#!/system/bin/sh" $1)" ] || sed -i "1i #!/system/bin/sh" $1
  local i; for i in "MODPATH" "LIBDIR" "MODID" "INFO" "MODDIR"; do
    case $i in
      "MODPATH") sed -i "1a $i=$NVBASE/modules/$MODID" $1;;
      "MODDIR") sed -i "1a $i=\${0%/*}" $1;;
      *) sed -i "1a $i=$(eval echo \$$i)" $1;;
    esac
  done
  [ "$1" == "$MODPATH/uninstall.sh" ] && return 0
  case $(basename $1) in
    post-fs-data.sh|service.sh) ;;
    *) cp_ch -n $1 $INPATH/$(basename $1) 0755;;
  esac
}

prop_process() {
  sed -i -e "/^#/d" -e "/^ *$/d" $1
  [ -f $MODPATH/system.prop ] || mktouch $MODPATH/system.prop
  while read LINE; do
    echo "$LINE" >> $MODPATH/system.prop
  done < $1
}

# Check for min/max api version
[ -z $MINAPI ] || { [ $API -lt $MINAPI ] && abort "! Your system API of $API is less than the minimum api of $MINAPI! Aborting!"; }
[ -z $MAXAPI ] || { [ $API -gt $MAXAPI ] && abort "! Your system API of $API is greater than the maximum api of $MAXAPI! Aborting!"; }

# Set variables
[ $API -lt 32 ] && DYNLIB=false
[ -z $DYNLIB ] && DYNLIB=false
[ -z $DEBUG ] && DEBUG=false
[ -e "$PERSISTDIR" ] && PERSISTMOD=$PERSISTDIR/magisk/$MODID
INFO=$NVBASE/modules/.$MODID-files
ORIGDIR="$MAGISKTMP/mirror"
if $DYNLIB; then
  LIBPATCH="\/vendor"
  LIBDIR=/system/vendor
else
  LIBPATCH="\/system"
  LIBDIR=/system
fi
if ! $BOOTMODE; then
  ui_print "- Only uninstall is supported in recovery"
  ui_print "  Uninstalling!"
  touch $MODPATH/remove
  [ -s $INFO ] && install_script $MODPATH/uninstall.sh || rm -f $INFO $MODPATH/uninstall.sh
  recovery_cleanup
  cleanup
  rm -rf $NVBASE/modules_update/$MODID $TMPDIR 2>/dev/null
  exit 0
fi

# Debug
if $DEBUG; then
  ui_print "- Debug mode"
  ui_print "  Module install log will include debug info"
  ui_print "  Be sure to save it after module install"
  set -x
fi

# Extract files
ui_print "- Extracting module files"
unzip -o "$ZIPFILE" -x 'META-INF/*' 'common/functions.sh' -d $MODPATH >&2
[ -f "$MODPATH/common/addon.tar.xz" ] && tar -xf $MODPATH/common/addon.tar.xz -C $MODPATH/common 2>/dev/null

# Run addons
if [ "$(ls -A $MODPATH/common/addon/*/install.sh 2>/dev/null)" ]; then
  ui_print " "; ui_print "- Running Addons -"
  for i in $MODPATH/common/addon/*/install.sh; do
    ui_print "  Running $(echo $i | sed -r "s|$MODPATH/common/addon/(.*)/install.sh|\1|")..."
    . $i
  done
fi

# Remove files outside of module directory
ui_print "- Removing old files"

if [ -f $INFO ]; then
  while read LINE; do
    if [ "$(echo -n $LINE | tail -c 1)" == "~" ]; then
      continue
    elif [ -f "$LINE~" ]; then
      mv -f $LINE~ $LINE
    else
      rm -f $LINE
      while true; do
        LINE=$(dirname $LINE)
        [ "$(ls -A $LINE 2>/dev/null)" ] && break 1 || rm -rf $LINE
      done
    fi
  done < $INFO
  rm -f $INFO
fi

### Install
ui_print "- Installing"

[ -f "$MODPATH/common/install.sh" ] && . $MODPATH/common/install.sh

ui_print "   Installing for $ARCH SDK $API device..."
# Remove comments from files and place them, add blank line to end if not already present
for i in $(find $MODPATH -type f -name "*.sh" -o -name "*.prop" -o -name "*.rule"); do
  [ -f $i ] && { sed -i -e "/^#/d" -e "/^ *$/d" $i; [ "$(tail -1 $i)" ] && echo "" >> $i; } || continue
  case $i in
    "$MODPATH/service.sh") install_script -l $i;;
    "$MODPATH/post-fs-data.sh") install_script -p $i;;
    "$MODPATH/uninstall.sh") if [ -s $INFO ] || [ "$(head -n1 $MODPATH/uninstall.sh)" != "# Don't modify anything after this" ]; then
                               install_script $MODPATH/uninstall.sh
                             else
                               rm -f $INFO $MODPATH/uninstall.sh
                             fi;;
  esac
done

$IS64BIT || for i in $(find $MODPATH/system -type d -name "lib64"); do rm -rf $i 2>/dev/null; done  
[ -d "/system/priv-app" ] || mv -f $MODPATH/system/priv-app $MODPATH/system/app 2>/dev/null
[ -d "/system/xbin" ] || mv -f $MODPATH/system/xbin $MODPATH/system/bin 2>/dev/null
if $DYNLIB; then
  for FILE in $(find $MODPATH/system/lib* -type f 2>/dev/null | sed "s|$MODPATH/system/||"); do
    [ -s $MODPATH/system/$FILE ] || continue
    case $FILE in
      lib*/modules/*) continue;;
    esac
    mkdir -p $(dirname $MODPATH/system/vendor/$FILE)
    mv -f $MODPATH/system/$FILE $MODPATH/system/vendor/$FILE
    [ "$(ls -A `dirname $MODPATH/system/$FILE`)" ] || rm -rf `dirname $MODPATH/system/$FILE`
  done
  # Delete empty lib folders (busybox find doesn't have this capability)
  toybox find $MODPATH/system/lib* -type d -empty -delete >/dev/null 2>&1
fi

# Set permissions
ui_print " "
ui_print "▌Setting Permissions"
set_perm_recursive $MODPATH 0 0 0755 0644
if [ -d $MODPATH/system/vendor ]; then
  set_perm_recursive $MODPATH/system/vendor 0 0 0755 0644 u:object_r:vendor_file:s0
  [ -d $MODPATH/system/vendor/app ] && set_perm_recursive $MODPATH/system/vendor/app 0 0 0755 0644 u:object_r:vendor_app_file:s0
  [ -d $MODPATH/system/vendor/etc ] && set_perm_recursive $MODPATH/system/vendor/etc 0 0 0755 0644 u:object_r:vendor_configs_file:s0
  [ -d $MODPATH/system/vendor/overlay ] && set_perm_recursive $MODPATH/system/vendor/overlay 0 0 0755 0644 u:object_r:vendor_overlay_file:s0
  for FILE in $(find $MODPATH/system/vendor -type f -name *".apk"); do
    [ -f $FILE ] && chcon u:object_r:vendor_app_file:s0 $FILE
  done
fi
set_permissions

# Complete install
cleanup
